proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1330 @@
+
+      In code::
+
+        interpretExpItem('module' 'qualified' modids expspec, ambient_env)
+          = let X' = interpretExpSpec(expspec, strip(modids, ambient_env)) in
+            qualify(modids, X')
+        interpretExpItem('module' 'qualified' modids, ambient_env)
+          = qualify(modids, strip(modids, ambient_env))
+
+   #. For an export item like ``module modids [ exports ]``:
+
+      The ambient, in-scope QEnv is ``E``. Define a new QEnv ``E'``
+      as follows::
+
+        E' = { qual ↦ orig | qual ↦ orig ∈ E
+                           , modids.qual ↦ orig ∈ E }
+
+      If ``exports`` is specified, let ``X`` be the interpretation of ``exports``
+      with respect to ``E'``.
+
+      If ``exports`` is not specified, let ``X = E'``.
+
+      This export item exports the environment ``X``.
+
+      In code::
+
+        interpretExpItem('module' modids maybe_exports, ambient_env)
+          = let E' = { qual ↦ orig | qual ↦ orig ∈ ambient_env
+                                   , modids.qual ↦ orig ∈ ambient_env } in
+            case maybe_exports of
+              Nothing -> E'
+              Just expspec -> interpretExpSpec(expspec, E')
+
+      Other than the new possibility of writing ``exports``, this is not a change
+      from current behavior: a ``module modids`` export exports all identifiers
+      in scope both with and without the ``modids`` qualification.
+
+      One non-backward-compatible alternative below is not to require that
+      names be in scope without qualification; this would make this unqualified
+      case more similar to the qualified case.
+
+      (This is still slightly backward-incompatible in an `obscure scenario`_.)
+
+   #. If there is no export list given for a module, all entities declared in that
+      module (and only those) are included in the export QEnv. This notion can
+      be made formal as follows::
+
+        X = { qual ↦ orig | qual ↦ orig ∈ ambient_env
+                          , orig = (MODIDS, _) }
+
+      where ``MODIDS`` is the fully qualified name of the current module.
+
+      This export ``X`` may contain overlapping entries. Here is an example::
+
+        module Top where
+          module A where
+            x = 5
+          module B where
+            x = True
+
+      In this example, the ambient QEnv for ``Top`` includes mappings
+      ``{ A.x ↦ (Top, A.x), x ↦ (Top, A.x), B.x ↦ (Top, B.x), x ↦ (Top, B.x) }``.
+      This QEnv contains two mappings for ``x``. By itself, this does not cause
+      a problem: it is only a usage site of ``x`` that would cause an ambiguity
+      error (as usual). However, since ``Top`` does not have an export list,
+      we would normally export all of the ambient QEnv in ``Top``, causing trouble
+      for importing modules.
+
+      We thus exclude from the default export list any qualified name that is mapped to multiple
+      original names. More formally::
```

Comment:
Will this cause hard to debug errors, when someone writing `import Top; … x … `  get’s a ”`x` not in scope” error?

Do you envision that GHC will remember that `Top` had `x` defined, but couldn’t export it, and give a helpful hint in that case?

Reply:
Good point. I will add a warning for this case, which really is a bit smelly (and deserving of a warning).